File: Structural/Lab7/C#/Entities/Room.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.AccessControl;
using System.Text;
using System.Threading.Tasks;

namespace C_.Entities
{
    public class Room : RealEstate
    {
        public string HeatingType { get; set; }
        public Room():base("Dungeon", "tra ta ta", 50, 1000, "New"){
        }
        public Room(string address, string description, double area, double price, string status) :
            base(address, description, area, price, status)
        {   
        }

        public override string ToString(){
            return "Room:\n"+ base.ToString();
        }

        public override void Rent(Customer customer, DateTime startDate, DateTime endDate){
            if(IsRentable)
                Console.WriteLine($"Customer {customer.NameAndSurname} rented room on address {Address} from {startDate} to {endDate}.");
        }

        public override void Sell(Customer customer){
            if(IsSellable)
                Console.WriteLine($"Customer {customer.NameAndSurname} bought room on address {Address}.");
        }
        
        public override void SetEstateToBuy(Customer customer, double price){
            if(IsBuyable){
                Console.WriteLine($"Customer {customer.NameAndSurname} put up for sale room on address {Address} with pice(USD) {price}.");
                PriceInUSD = price;
                IsBuyable = false;
            }
        }
        public override Room Clone()
        {
            var clone = (Room)MemberwiseClone();
            clone.Realtor = Realtor == null ? Realtor : new Realtor(Realtor);
            return clone;
        }

        public override Room ShallowCopy()
        {
            return (Room)MemberwiseClone();
        }

        public override void Reset()
        {
            base.Reset();
            HeatingType = "";
        }
    }
}

File: Structural/Lab7/Typescript/src/Entities/room.ts
import { RealEstate } from './real_estate';
import { Customer } from './customer';
import { Realtor } from './realtor';
export class Room extends RealEstate {
    public heatingType: string;

    constructor();
    constructor(address: string, description: string, area: number, price: number, status: string);
    constructor(address?: string, description?: string, area?: number, price?: number, status?: string) {
        if (address && description && area && price && status) {
            super(address, description, area, price, status);
        } else {
            super("Dungeon", "tra ta ta", 50, 1000, "New");
        }
    }

    public toString(): string {
        return "Room:\n" + super.toString();
    }

    public rent(customer: Customer, startDate: Date, endDate: Date): void {
        if (this.isRentable)
            console.log(`Customer ${customer.nameAndSurname} rented room on address ${this.address} from ${startDate} to ${endDate}.`);
    }

    public sell(customer: Customer): void {
        if (this.isSellable)
            console.log(`Customer ${customer.nameAndSurname} bought room on address ${this.address}.`);
    }

    public setEstateToBuy(customer: Customer, price: number): void {
        if (this.isBuyable) {
            console.log(`Customer ${customer.nameAndSurname} put up for sale room on address ${this.address} with price(USD) ${price}.`);
            this.priceInUSD = price;
            this.isBuyable = false;
        }
    }

    public clone(): Room {
        const clone = Object.assign(Object.create(Object.getPrototypeOf(this)), this);
        clone.realtor = this.realtor ? Realtor.newRealtorFromObject(this.realtor) : null;
        return clone;
    }

    public shallowCopy(): Room {
        return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
    }

    public reset(): void {
        super.reset();
        this.heatingType = "";
    }
}